crypto/types/types.proto

```diff
@@ -0,0 +1,52 @@
+syntax = "proto3";
+package cosmos_sdk.crypto.v1;
+
+import "third_party/proto/gogoproto/gogo.proto";
+import "google/protobuf/any.proto";
+
+option go_package = "github.com/cosmos/cosmos-sdk/crypto/types";
+
+// PublicKey specifies a public key
+message PublicKey {
+    // sum specifies which type of public key is wrapped
+    oneof sum {
+        bytes secp256k1 = 1;
+        bytes ed25519 = 2;
+        bytes sr25519 = 3;
+        PubKeyMultisigThreshold multisig = 4;
+        bytes secp256r1 = 5;
```

Comment:
I think only having keys that are tested in the sdk be added here by default (tendermint/secp256k1). Sr25519, ed25519 & secp256r1 are not tested in the application. If the developer would like to add them they would be able to do this via the any_pubkey.

Reply:
From the updated [ADR 020](https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/adr-020-protobuf-transaction-encoding.md#transactions):

> Because we are aiming for a flexible, extensible cross-chain transaction format, new transaction processing options should be added to TxBody as soon those use cases are discovered, even if they can't be implemented yet.
> 
> Because there is coordination overhead in this, TxBody includes an extension_options field which can be used for any transaction processing options that are not already covered. App developers should, nevertheless, attempt to upstream important improvements to Tx.

So my thought is that we take the same philosophy with `PublicKey`. We attempt to cover all known algorithms that people might use, but applications need to provide a `PublicKeyDecoder` to actually use them

Reply:
@marbar3778 I've updated the text in the ADR here in this PR to explain this.